README.md
<p>

  [![CI](https://github.com/PyDrocsid/MorpheusHelper/actions/workflows/ci.yml/badge.svg)](https://github.com/PyDrocsid/MorpheusHelper/actions/workflows/ci.yml)
  [![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
  [![DockerHub - MorpheusHelper](https://img.shields.io/docker/pulls/defelo/morpheushelper?style=flat-square&label=DockerHub%20-%20MorpheusHelper)](https://hub.docker.com/r/defelo/morpheushelper)
  [![Discord](https://img.shields.io/discord/637234990404599809.svg?label=Discord&logo=discord&logoColor=ffffff&color=7389D8)](https://discord.gg/zcmkZVC)

</p>

# MorpheusHelper
Bot for the [Discord Server of The Morpheus Tutorials](https://discord.gg/themorpheus)

## Development
### Prerequisites
- [Python](https://python.org/) >=3.9
- [Pipenv](https://github.com/pypa/pipenv/)
- [Git](https://git-scm.com/)
- [Docker](https://www.docker.com/) (recommended)
- [docker-compose](https://docs.docker.com/compose/) (recommended)
- [PyCharm Community/Professional](https://www.jetbrains.com/pycharm/) (recommended)

### Clone the repository

#### SSH (recommended)
```
git clone --recursive git@github.com:PyDrocsid/MorpheusHelper.git
```

#### HTTPS
```
git clone --recursive https://github.com/PyDrocsid/MorpheusHelper.git
```

### Setup dependencies

After you have cloned the repository, you should create a virtual environment and install all dependencies. For this you can use the following command:

```
pipenv install --dev
```

### Environment variables
To set the required environment variables it is necessary to create a file named `.env` in the root directory (there is a template for this file in [`morpheushelper.env`](morpheushelper.env)). If you need a token, generate one by following these instructions: [Creating a Bot Account](https://discordpy.readthedocs.io/en/latest/discord.html) (Note you need to enable the options under `Privileged Gateway Intents`)

### Project structure

```
Project
├── bot
│  ├── cogs
│  │  ├── custom
│  │  │  ├── <cog>
│  │  │  │  ├── translations
│  │  │  │  │  └── en.yml
│  │  │  │  ├── __init__.py
│  │  │  │  ├── api.py
│  │  │  │  ├── cog.py
│  │  │  │  ├── colors.py
│  │  │  │  ├── models.py
│  │  │  │  ├── permissions.py
│  │  │  │  └── settings.py
│  │  │  ├── contributor.py
│  │  │  ├── pubsub.py
│  │  │  └── translations.py
│  │  └── library (submodule)
│  │     ├── <category>
│  │     │  └── <cog>
│  │     │     ├── translations
│  │     │     │  └── en.yml
│  │     │     ├── __init__.py
│  │     │     ├── api.py
│  │     │     ├── cog.py
│  │     │     ├── colors.py
│  │     │     ├── documentation.md
│  │     │     ├── models.py
│  │     │     ├── permissions.py
│  │     │     └── settings.py
│  │     ├── contributor.py
│  │     ├── pubsub.py
│  │     └── translations.py
│  ├── bot.py
│  └── morpheushelper.py
└── config.yml
```

### PyCharm configuration

- Open PyCharm and go to `Settings` ➔ `Project: MorpheusHelper` ➔ `Python Interpreter`
- Open the menu `Python Interpreter` and click on `Show All...`
- Click on the plus symbol
- Click on `Pipenv Environment`
- Select `Python 3.9` as `Base interpreter`
- Confirm with `OK`
- Change the working directory to root path  ➔ `Edit Configurations`  ➔ `Working directory`


Finally, please remember to mark the `bot` directory as `Sources Root` (right click on `bot` ➔ `Mark Directory as` ➔ `Sources Root`).


## Installation instructions

### Using Docker
```bash
# clone git repository and cd into it
git clone --recursive https://github.com/PyDrocsid/MorpheusHelper.git
cd MorpheusHelper

# build docker image
sudo docker build -t pydrocsid/morpheushelper .

# adjust the docker-compose.yml and create the .env file
cp morpheushelper.env .env
vim .env
vim docker-compose.yml

# start redis, database and bot using docker-compose
sudo docker-compose up -d
```

### Local installation
```bash
# install pipenv
pip install pipenv

# create virtual environment and install requirements
pipenv install

# start the bot
pipenv run bot
```

### Environment variables
|    Variable Name    |                                     Description                                      |  Default Value   |
|:--------------------|:-------------------------------------------------------------------------------------|:-----------------|
| TOKEN               | Discord Bot Token                                                                    |                  |
| GITHUB_TOKEN        | GitHub Personal Access Token (PAT) with public access                                |                  |
| LOG_LEVEL           | one of `DEBUG`, `INFO`, `WARNING`, `ERROR`, `CRITICAL`                               | `INFO`           |
|                     |                                                                                      |                  |
| DB_DRIVER           | Name of the SQL connection driver                                                    | `mysql+aiomysql` |
| DB_HOST             | Hostname of the database server                                                      | `localhost`      |
| DB_PORT             | Port on which the database server is running                                         | `3306`           |
| DB_DATABASE         | Name of the database you want to use                                                 | `bot`            |
| DB_USER             | Username for the database account                                                    | `bot`            |
| DB_PASSWORD         | Password for the database account                                                    | `bot`            |
| POOL_RECYCLE        | Number of seconds between db connection recycling                                    | `300`            |
| POOL_SIZE           | Size of the connection pool                                                          | `20`             |
| MAX_OVERFLOW        | The maximum overflow size of the connection pool                                     | `20`             |
| SQL_SHOW_STATEMENTS | whether SQL queries should be logged                                                 | `False`          |
|                     |                                                                                      |                  |
| REDIS_HOST          | Hostname of the redis server                                                         | `redis`          |
| REDIS_PORT          | Port on which the redis server is running                                            | `6379`           |
| REDIS_DB            | Index of the redis database you want to use                                          | `0`              |
|                     |                                                                                      |                  |
| CACHE_TTL           | Time to live of redis cache in seconds                                               | `28800`          |
| RESPONSE_LINK_TTL   | Time to live of links between command messages and the resulting bot responses       | `7200`           |
|                     |                                                                                      |                  |
| REPLY               | wether to use the reply feature when responding to commands                          | `True`           |
| MENTION_AUTHOR      | wether to mention the author when the reply feature is being used                    | `True`           |
|                     |                                                                                      |                  |
| SENTRY_DSN          | [Optional] Sentry DSN for logging                                                    |                  |
| OWNER_ID            | [Optional] Discord User ID of the person who should recieve status information.      |                  |
| DISABLED_COGS       | [Optional] Cogs you'd like to disable.                                               |                  |
|                     |                                                                                      |                  |
| AOC_SESSION         | [Optional] Session cookie of the [Advent of Code Website](https://adventofcode.com/) |                  |
| AOC_REFRESH_INTERVAL| [Optional] Number of seconds after which the aoc cache is invalidated                | `900`            |

# MorpheusHelper Technical Documentation

**Last updated:** 2022-01-20\
_Document generation aided by **Documatic**_

Automatic Documentation

* [Introduction](#introduction)
* [Code Overview](#code-overview)

## Introduction

This is a technical document detailing
        at a high-level
        what MorpheusHelper does, how it operates,
        and how it is built.

The outline of this document was generated
        by **Documatic**.
<!---Documatic-section-group: arch-start--->


## Project Architecture


<!---Documatic-section-arch: container-start--->
The project uses Docker to build a maintainable environment for the project, using python:3.10-alpine as a base image. During build, the container expects ARG variables PIPENV_NOSPIN=true, PIPENV_VENV_IN_PROJECT=true.


<!---Documatic-section-arch: container-end--->

<!---Documatic-section-group: arch-end--->

<!---Documatic-section-group: helloworld-start--->


## Code Overview

The codebase has a flat structure.
<!---Documatic-section-helloworld: setup-start--->

No exact compatable Python version has been detected.
Versions below 3.8 are compatable.

Install from pypi with `pip install main`.



<!---Documatic-section-helloworld: setup-end--->

<!---Documatic-section-helloworld: entrypoints-start--->


<!---Documatic-section-helloworld: entrypoints-end--->

<!---Documatic-section-group: concept-start--->
## Concepts
<!---Documatic-section-group: concept-end--->

<!---Documatic-section-group: helloworld-end--->

<!---Documatic-section-group: dev-start--->


## Developers
<!---Documatic-section-dev: setup-start--->





<!---Documatic-section-dev: setup-end--->

<!---Documatic-section-dev: ci-start--->
The project uses GitHub Actions for CI/CD.

| CI File | Purpose |
|:----|:----|
| docker_clean | Executes on delete for any branch |
| ci | Executes on push for any branch, pull_request for any branch |


<!---Documatic-section-dev: ci-end--->

<!---Documatic-section-group: dev-end--->


CHANGELOG.md
# Changelog

Based on KeepAChangelog.
Generated by **Documatic.**

## Unreleased

### Added

* Confirmation countdown

### Changed

* Piston integration to v2

### Fixed

* Pagination timeout

### Removed

* Vc reaction handler

## 3.3.0 - 2022-01-07

### Added

* Library submodule
* Docker healthcheck
* Default collation for mysql
* Username to user notes embed
* Basic thread utils
* W503 to .flake8
* User notes to ulog

### Changed

* Library to pycord
* Python to 3.10
* Unicode handling in reactionrole

### Fixed

* Confirmations
* Broken urls
* Versions in ci.yml
* Replies
* Config.yml

### Removed

* Utc hints

## 3.2.1 - 2021-12-22

### Changed

* Python to 3.10
* Aiomysql

### Fixed

* Linter errors
* Import
* Ci

## 3.2.0 - 2021-11-15

### Added

* Aoc default permissions
* Docker healthcheck


## 3.1.6 - 2021-11-13

### Added

* Library submodule

### Fixed

* Broken urls

## 3.1.5 - 2021-10-31

### Changed

* Btp role assignment commands

## 3.1.4 - 2021-10-13

### Added

* All datetime objects timezone aware

### Changed

* Library to pycord branch
* Submodules
* Dependencies
* Library

### Fixed

* Datetime issues in inactivity cog
* Utcdatetime
* Avatar_url and icon_url
* Sudo clear_cache command

## 3.1.3 - 2021-09-05

### Changed

* Vc rename command

## 3.1.0 - 2021-08-29

### Changed

* Dependencies
* Cogs
* Pipfile

### Fixed

* Internal errors in reactionrole
* Durationconverter
* Polloption emoji parser
* Rr add command
* Reddit translations
* Inactivity scan command
* Inactive duration limits

### Removed

* Library submodule

## 3.0.0 - 2021-08-26

### Added

* Github actions workflow
* Github environment
* Wemake to ci
* Flake8 plugins and fixed codestyle
* Gitsubmodule to dependabot
* .flake8 and fixed ci
* Missing env vars to morpheushelper.env
* Pydrocsid library submodule
* Docker deployment
* Fine-grained permissions
* Clear_cache command
* Redis to spam detection
* Redis to python documentation cog
* Redis to mediaonly channels
* Submodule remotes relative
* Autorole cog
* Reverse flag to reactionrole
* Role notifications cog
* User info cog
* Inactivity cog
* Badges to readme.md
* Error messages to roles add and remove commands
* Support for modcommanderror and embed author
* Team yesno command
* Pagination
* Cooldown to pagination
* Disable_pagination env var
* Repeat_footer enforcement when using pagination
* Custom commands cog
* Pagination env vars
* Github id optional for contributors
* Artifact retention to ci
* Custom_commands volume to docker-compose.yml
* Custom command permissions
* Suggest_role_color command
* Optional confirmation requirement to custom commands
* Permission_error event handler
* Permission checks
* New dynamic voice channels
* Vc help command
* Reactions to bot messages
* Vc rename command
* Status icon to roles list command
* Status icon to inactive command
* Message clear command
* Intel codenames
* Attachments to remindme
* Delta to list of contributors
* Language and version to run code output embed
* Perma roles
* Discord bot token info command
* Discord bot token deleter
* Merge me workflow
* Cog documentation to readme
* Docker clean workflow
* S311 to .flake8
* Random vc name list mechanism
* Documentation to cogs command
* Cog documentation to help command
* Documentation for suggest_role_color command
* Encode command
* Autoclearcog
* Emojis to command list in help embed
* New custom commands
* Cc permission alerts
* Warning to cc channel parameter
* Permission check to cc e channel command
* Discord badge to readme
* Default permission levels to config (fixes #246)

### Changed

* Cog library structure and added basic logging
* Cogs into library submodule
* Pydrocsid to v2.0.0.dev1
* Bot replies
* Invite and mediaonly alerts into alert channel
* And improved readme.md
* Cogs and library
* Docker push job in ci
* Cog library submodule
* Role assignment checks
* Ci and updated submodules
* Library submodule
* Codestyle job in ci
* Mod embeds
* Team role names
* Library and submodules
* Gh repo config
* Function names of btp commands
* Cog library
* Library to aa31a22 to test new send_long_embed implementation
* Codeblock and metaquestion info commands into custom commands
* Dependencies and submodules
* Inactive command
* Emoji for hidden channels
* Vc info command
* Vc d embed
* Voice handler delays
* Bot info embed
* Permission checks for role assignment commands
* Voice role links alerts
* Voice channel cog
* Suggest_role_color command
* Verify too_soon error message

### Fixed

* Alert channel set command
* Dockerfile
* Dependencies
* Ci for pull requests from forks
* Codestyle
* Ci
* Docker image
* Pydrocsid version
* Name of runcodecog
* Init_join_log
* Import in morpheushelper.py
* Import of modcommanderror and updated library
* Option limit in polls
* Split_lines function
* Btp translations
* On_command_error
* Library imports
* Sudo cog
* Status description in heartbeat cog
* Dynamic voice channels
* Sf command
* Custom command paths
* Vc show command
* Quick rejoin after kick issue
* Role links in dynamic voice channels
* Text channel permissions
* Vc remove command
* Member order in info embed
* Contributor info command
* And improved user notes
* Contributor command
* Permission error in leave event handler
* Hide/show reactions
* Permission errors (again)
* Connect permission overwrites
* Roles perma_list command
* Logger edit distance bug
* And improved mediaonly cog
* Verify command
* Vc permission overwrite bug
* Cc empty content error
* Cc embed errors
* Cc empty embed error
* And improved custom commands
* Cc on_ready
* Roles config commands
* Internal errors in reactionrole
* Durationconverter
* Polloption emoji parser
* Rr add command
* Reddit translations
* Inactivity scan command
* Inactive duration limits

### Removed

* Useless get_last_autokick pubsub channel
* Library submodule

## 2.3.0 - 2021-03-16

### Added

* Documentation for exploit workaround

database.py
from asyncio import BoundedSemaphore
from os import environ as env
from typing import TypeVar, Optional, Union, Iterable, Type, List

from sqlalchemy import create_engine
from sqlalchemy.engine import Engine
from sqlalchemy.ext.declarative import DeclarativeMeta, declarative_base
from sqlalchemy.orm import sessionmaker, scoped_session, Query, Session

import async_thread

T = TypeVar("T")


class DB:
    def __init__(self, hostname, port, database, username, password):
        self.engine: Engine = create_engine(
            f"mysql+pymysql://{username}:{password}@{hostname}:{port}/{database}?charset=utf8mb4",
            pool_pre_ping=True,
            pool_recycle=300,
            pool_size=10,
            max_overflow=20,
        )

        self._SessionFactory: sessionmaker = sessionmaker(bind=self.engine, expire_on_commit=False)
        self._Session = scoped_session(self._SessionFactory)
        self.Base: DeclarativeMeta = declarative_base()

    def create_tables(self):
        self.Base.metadata.create_all(bind=db.engine)

    def close(self):
        self._Session.remove()

    def add(self, obj: T):
        self.session.add(obj)

    def delete(self, obj: T):
        self.session.delete(obj)

    def query(self, model: Type[T], **kwargs) -> Union[Query, Iterable[T]]:
        return self.session.query(model).filter_by(**kwargs)

    def all(self, model: Type[T], **kwargs) -> List[T]:
        return self.query(model, **kwargs).all()

    def first(self, model: Type[T], **kwargs) -> Optional[T]:
        return self.query(model, **kwargs).first()

    def count(self, model: Type[T], **kwargs) -> int:
        return self.query(model, **kwargs).count()

    def get(self, model: Type[T], primary_key) -> Optional[T]:
        return self.session.query(model).get(primary_key)

    @property
    def session(self) -> Session:
        return self._Session()


thread_semaphore = BoundedSemaphore(5)


async def run_in_thread(function, *args, **kwargs):
    async with thread_semaphore:

        def inner():
            try:
                out = function(*args, **kwargs)
                db.session.commit()
            finally:
                db.close()
            return out

        return await async_thread.run_in_thread(inner)


db: DB = DB(
    hostname=env.get("DB_HOST", "localhost"),
    port=env.get("DB_PORT", 3306),
    database=env.get("DB_DATABASE", "morpheushelper"),
    username=env.get("DB_USER", "morpheushelper"),
    password=env.get("DB_PASSWORD", "morpheushelper"),
)
